### DISCUSSION TITLE
Character creation works, connecting does not.

### DISCUSSION THREAD

**Discussion by TheLastCylon** [evennia/evennia#2529] [Evennia Questions] [answered]
Greetings all,

I've been playing with Evennia for a few weeks now, and all looks great.

Thing is, today I tried to create a character that is not a super user. All character creation is still the default process.
So I do the following:
`create Robby <some password>`
The character creation works fine, and I get the message that I can now do:
`You can now log with the command 'connect Robby <your password>'.`

However, when I do execute the command `connect Robby <the password from before>` I get the message:
`The Character does not exist.`

I'm using the latest from master branch, so ... not really sure how to proceed without hacking code in `accounts.py`.

Am I missing something?

**Comment by Griatch** [comment]
Hi @TheLastCylon 
This should work. I just tested this process in master branch, both for a telnet client and with the webclient - I even tested naming my character `Robby` just to be sure ... and it works fine here. 

Some thing to try:

- Does your password contain spaces? In that case you may need to use quotes. 
- My first guess would be that you have introduced some error in your account class. If so you should be seeing a traceback in your log. Check your log with `evennia --log` (exit the log with Ctrl-C later) and see if something happens when you create the account.
- Check if the account actually exists or not. In-game, do `py evennia.search_account("Robby")` and see if you find it. If not, something went fishy in creation (did you modify the `create` command?). If you find it, something went wrong in the login-search (did you change the `connect` command, or change permissions on the account?)

**Reply by TheLastCylon** [reply]
Thanks for the reply Griatch,

To answer your questions:
1. None of the passwords I've used contained spaces.
2. Here's a log excerpt from when I do a create, and then a connect:
```
21-09-10 16:01:07 Evennia Server successfully started.
21-09-10 16:02:09 [SS] Password successfully changed for Newuser(account None).
21-09-10 16:02:09 [SS] Account Created: Newuser(account 8) (IP: 127.0.0.1).
21-09-10 16:02:09 [EE] substring not found
21-09-10 16:02:40 [SS] Authentication Success: Newuser(account 8) (IP: 127.0.0.1).
21-09-10 16:02:40 [..] Logged in: Newuser(account 8) 127.0.0.1 (1 session(s) total)
```
The `[EE] substring not found` line looks suspect, but it's not causing issues in the client.
The output I get in the client is as described in my original post.

Notes:
- I've not altered any of the code in the evennia directory.
- I've not altered `accounts.py` for the game directory either.
- Neither the create nor the connect commands have been overridden in any way.
- `py evennia.search_account("Robby")` produces no output, but it does not produce any output for my super user either, and I can log in with that account just fine.

I'm still playing around, so I'm using the default SQLite DB at the moment. Started digging in there with a few queries and found the accounts in the `accounts_accountdb` table, I can actually see the `username` for all the users I've tried to create.
i.e. The query:
```
select id, last_login, username, is_staff, db_key, db_typeclass_path, db_cmdset_storage
from accounts_accountdb
where username = 'Newuser';
```
Results in the following:
```
8, 2021-09-10 16:02:40.881485, Newuser , 0, #8, typeclasses.accounts.Account, commands.default_cmdsets.AccountCmdSet
```

I can also find the user object in the `objects_objectdb` table.
Here's what's a bit suspect though.

In `accounts_accountdb` the `db_key` column for this user is `#8`
But in `objects_objectdb` the objects id in the database is `24`

Are these related? Should they not have the same number?

Again, thanks for the response, much appreciated. :)

**Reply by TheLastCylon** [reply]
Am I correct if I say the `id` and the `db_key` in the `accounts_accountdb` table are the same, because `db_key` is a textual reference to the value of `id`, if so, my last questions are answered :)

**Reply by Griatch** [reply]
`accounts_accountdb`, the `db_key` is the name of your account, so it should be `Robby` and not `#8` ...  This seems wrong. 

The link between the account and object is the `objects_objectdb` table, `db_account_id` is set to the the `id` of your account, but _only when you are logged in and are puppeting them_. What tells the login mechanism to puppet is instead the `_last_puppet` Attribute on the account.

**Comment by TheLastCylon** [comment, marked as answer]
On a side note:

I've just done the test on a newly initialized game as well. All seems to be working there, so it must be something I've coded in my original attempt.

Thanks for the assist, but I think I'm going to have to sift through the stuff I've changed and see if there's something there, that's causing this.

**Reply by Griatch** [reply]
That's quite an important side note :D It looks like there may have been a confusion between `key` and `dbref` (see my other reply), since you end up with `account_acccountdb.db_key` being `#8` rather than it being the string "Robby". So that's a good place to start looking.

**Reply by TheLastCylon** [reply]
Actually, I think I discovered why that was happening:

I was including `CmdCreate` in my `default_cmdsets.py`. This was while I was experimenting with various ways to ensure a command prompt and gmcp message, every time the user enters a command or receives a text message of some kind.

I eventually settled on just overriding the `msg` method for the `Character` class, I'm a little concerned about performance with this approach, but considering the average mud barely has 100 users active at any one time, that really should not be an issue. :)

I'm going to go ahead and mark this as answered, as that was ... my-bad.

**Reply by Griatch** [reply]
Glad you figured it out!
